Main.py

```diff
@@ -29,21 +29,36 @@ def run():
         still_in_cycle = 1 # 1
         while still_in_cycle:
 
-            t += MINIMUM_TIME_RESOLUTION # Possible speed up? find biggest time increment...
-
             # Check lag
             if lag:
+                time_step = MINIMUM_TIME_RESOLUTION
                 lagging = (sim_env['lag_time'][:nr_alive] > 0)
                 lag = any(lagging)
-                sim_env['lag_time'][lagging] -= MINIMUM_TIME_RESOLUTION
-                sim_env['next_divition'][lagging] += MINIMUM_TIME_RESOLUTION
+                sim_env['lag_time'][lagging] -= time_step
+                sim_env['next_divition'][lagging] += time_step
+            else:
+                if time_step != MINIMUM_TIME_RESOLUTION or np.random.random() < A_MYSTICAL_SPEED_UP_PARAMETER:
+                        soonest_division = np.min([sim_env['next_divition'][:nr_alive]])
+                        time_step = np.max([MINIMUM_TIME_RESOLUTION, soonest_division - t]) # find biggest time increment
+
+            t += time_step
 
             ## Divide
             to_divide = (sim_env['next_divition'][:nr_alive] <= t)
             nr_divide = np.sum(to_divide)
             if nr_divide:
                 divide_index = np.nonzero(to_divide)[0]
                 to_birth = np.arange(nr_alive,(nr_alive + nr_divide))
+
+                if nr_alive + len(to_birth) > MAXIMUM_NR_AGENTS:
+                    nr_divide_reduced = MAXIMUM_NR_AGENTS - nr_alive
+                    to_birth = to_birth[nr_divide_reduced]
```

Comment:
This should be `to_birth[:nr_divide_reduced]`, I think, else `to_birth` ought to be a number after this line. If intended, it would be nice to have a different name so that it is less confusing.